lib/web/command.go

```diff
@@ -20,20 +20,614 @@
 )
 
+const CommandResultType = "COMMAND_RESULT"
+
+type CommandRequest struct {
+	// Command is the command to be executed on all nodes.
+	Command string `json:"command"`
+	// Login is a Linux username to connect as.
+	Login string `json:"login"`
+	// Query is the predicate query to filter nodes where the command will be executed.
+	Query string `json:"query"`
+	// ConversationID is the conversation context that was used to execute the command.
+	ConversationID string `json:"conversation_id"`
+	// ExecutionID is a unique ID used to identify the command execution.
+	ExecutionID string `json:"execution_id"`
+}
+
+func (c *CommandRequest) Check() error {
+	if c.Command == "" {
+		return trace.Errorf("missing command")
+	}
+
+	if c.Login == "" {
+		return trace.Errorf("missing login")
+	}
+
+	if c.ConversationID == "" {
+		return trace.Errorf("missing conversation ID")
+	}
+
+	if c.ExecutionID == "" {
+		return trace.Errorf("missing execution ID")
+	}
+
+	return nil
+}
+
+// CommandExecutionResult is a result of a command execution.
+type CommandExecutionResult struct {
+	SessionID string `json:"session_id"`
+}
+
+// executeCommand executes a command on all nodes that match the query.
 func (h *Handler) executeCommand(
 	w http.ResponseWriter,
 	r *http.Request,
 	_ httprouter.Params,
 	sessionCtx *SessionContext,
 	site reversetunnel.RemoteSite,
 ) (any, error) {
-	return nil, trace.NotImplemented("not implemented")
+	q := r.URL.Query()
+	params := q.Get("params")
+	if params == "" {
+		return nil, trace.BadParameter("missing params")
+	}
+	var req *CommandRequest
+	if err := json.Unmarshal([]byte(params), &req); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := req.Check(); err != nil {
+		return nil, trace.BadParameter("invalid payload: %v", err)
+	}
+
+	clt, err := sessionCtx.GetUserClient(r.Context(), site)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	identity, err := createIdentityContext(req.Login, sessionCtx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ctx, err := h.cfg.SessionControl.AcquireSessionContext(r.Context(), identity, h.cfg.ProxyWebAddr.Addr, r.RemoteAddr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	authAccessPoint, err := site.CachingAccessPoint()
+	if err != nil {
+		h.log.WithError(err).Debug("Unable to get auth access point.")
+		return nil, trace.Wrap(err)
+	}
+
+	netConfig, err := authAccessPoint.GetClusterNetworkingConfig(ctx)
+	if err != nil {
+		h.log.WithError(err).Debug("Unable to fetch cluster networking config.")
+		return nil, trace.Wrap(err)
+	}
+
+	clusterName := site.GetName()
+
+	upgrader := websocket.Upgrader{
+		ReadBufferSize:  1024,
+		WriteBufferSize: 1024,
+		CheckOrigin:     func(r *http.Request) bool { return true },
+	}
+
+	ws, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		errMsg := "Error upgrading to websocket"
+		h.log.WithError(err).Error(errMsg)
+		http.Error(w, errMsg, http.StatusInternalServerError)
```

Comment:
Are we sure this works? I feel like I've seen some issues recently where you can't respond with normal HTTP errors after attempting the websocket upgrade.

Reply:
Copied from https://github.com/gravitational/teleport/blob/341340baeb5790b2ef93e6ffcc13755fa79bb3a4/lib/web/terminal.go#L296, so if it doesn't work at least it's consistent :sweat_smile: 
I think that connection upgrade should return a regular HTTP error, as WS is not created, but I'll double-check that.